File: laboratory/number_baseball.py
import random


def generate_random_number():
    return str(random.randint(100, 999))


def get_game_result(user_number_str, answer_number_str):
    strike = 0
    ball = 0

    # 3자리 숫자임을 가정해서 range(3)
    for i in range(3):
        # 정답과 일치하는 숫자가 있는 경우
        if user_number_str[i] in answer_number_str:
            if user_number_str[i] == answer_number_str[i]:
                # 정답과 일치하는 숫자가 위치도 일치하는 경우
                strike += 1
            else:
                # 위치가 일치하지 않는 경우
                ball += 1

    # 다 맞춘 경우
    if strike == 3:
        print('Yes! The secret number is "' + answer_number_str + '"! You have won!')
        print()
        return True

    # 다 맞추지 못한 경우
    print(str(strike), 'strike,', str(ball), 'ball.')
    print()
    print('==========' * 3)

    return False


def main():

    # 랜덤 숫자 생성
    answer = generate_random_number()

    # 시작 문구 표시
    print('S T A R T')
    print('==========' * 3)

    while True:
        print('Guess numbers (000 ~ 999):')
        # 유저 숫자 입력
        user_answer = input()

        # 겹치는 숫자 다 맞췄는지 결과 리턴
        game_result = get_game_result(user_answer, answer)

        # 겹치는 숫자 맞춘 경우는 다시 게임할 것인지 물어봄
        if game_result:
            user_input = input('Do you want to play again? (yes or no)')
            if user_input == 'no':
                break
            else:
                # 새 랜덤 숫자 생성
                answer = generate_random_number()


if __name__ == '__main__':
    main()

File: laboratory/number_baseball.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>

#define NUMBER_SIZE 3

int generateRandomNumber() {
	int randomNumber;

	// 랜덤 숫자 생성을 위한 씨드 설정
	srand(time(NULL));
	// 100 ~ 999 = 0 ~ 899 + 100
	randomNumber = rand() % 900 + 100;

	return randomNumber;
}

int getNumberFromUser() {
	int number;

	printf("Guess number (000 ~ 999): ");
	scanf("%d", &number);

	return number;
}

bool checkNumber(int inputNumber, int randomNumber) {
	int strikeCount, ballCount;
	int copiedRandomNumber = randomNumber;

	// 3자리 숫자임을 가정해서 숫자를 배열에 자릿수 별로 담아둠
	int inputNumberArray[NUMBER_SIZE];
	int randomNumberArray[NUMBER_SIZE];
	int i ;
	for (i = 0; i < NUMBER_SIZE; i++) {
		inputNumberArray[i] = inputNumber % 10;
		inputNumber = inputNumber / 10;

		randomNumberArray[i] = randomNumber % 10;
		randomNumber = randomNumber / 10;
	}

	for (i = 0; i < NUMBER_SIZE; i++) {
		// 자릿수와 숫자가 일치하면 스트라이크 1 증가
		if (inputNumberArray[i] == randomNumberArray[i]) {
			strikeCount += 1;
		} else {
			// 자릿수와 숫자가 일치하지 않으면 해당 숫자가 randomNumber 안에 있는지 체크함
			for (int j=0; j < 3; j++) {
				if (j != i && inputNumberArray[i] == randomNumberArray[j]) {
					ballCount += 1;
				}
			}
		}
	}

	if (strikeCount == 3) {
		// 3자리 숫자를 다 맞춘 경우
		printf("Correct, answer is %d\n", copiedRandomNumber);
	} else if (strikeCount > 0 || ballCount > 0) {
		// 일치하는 숫자가 1개라도 있는 경우
		printf("%d strike, %d ball\n", strikeCount, ballCount);
	} else {
		// 입력된 숫자가 랜덤 숫자에 하나도 포함되어있지 않은 경우
		printf("Out!\n");
	}

	return (strikeCount == 3);
}

int main(void) {
	char playYn;
	bool result;
	int inputNumber;
	int randomNumber = generateRandomNumber();

	printf("==================\n");
	printf("S T A R T\n");

	while(1) {
		inputNumber = getNumberFromUser();
		result = checkNumber(inputNumber, randomNumber);

		// 3 Strike인 경우 성공
		if (result) {
			printf("Play again? (y or n): ");
			scanf("%s", &playYn);

			if (playYn == 'n') {
				break;
			}

			// 새 랜덤 숫자를 생성
			randomNumber = generateRandomNumber();
		}
	}

    return 0;
}